### ISSUE TITLE
Unit Activation does not consider TimeoutStartSec

### ISSUE THREAD

**Issue by bernadinm** [systemd/systemd#6815]
### Submission type

 - Bug report

### systemd version the issue has been seen with

```bash
systemctl --version
systemd 229
+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ -LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD -IDN
```

### Used distribution

```bash
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=16.04
DISTRIB_CODENAME=xenial
DISTRIB_DESCRIPTION="Ubuntu 16.04 LTS"
NAME="Ubuntu"
VERSION="16.04 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
UBUNTU_CODENAME=xenial
```
### In case of bug report: Expected behaviour you didn't see

The systemd unit will stay activating forever.

### In case of bug report: Unexpected behaviour you saw

The systemd unit activation fails at 60 seconds every time regardless of the `TimeoutStartSec` option used. 

### In case of bug report: Steps to reproduce the problem

Just copy and paste this command into the terminal. No modification needed.

```bash
cat << 'EOF' | sudo tee /etc/systemd/system/ready-spartan-checker.service
[Unit]
Description=ready.spartan startup assert check
Documentation=https://www.freedesktop.org/software/systemd/man/systemd.unit.html

[Service]
SyslogIdentifier=ready-spartan-checker
ExecStartPre=/usr/bin/bash -c 'ping -c1 ready.spartan &>/dev/null; while [ $? -ne 0 ]; do printf "waiting for spartan to get ready... \n"; sleep 5; ping -c1 ready.spartan &>/dev/null; done; printf "Spartan is now ready."'
ExecStart=/bin/sleep 1
TimeoutStartSec=infinity
RemainAfterExit=yes
EOF

cat << 'EOF' | sudo tee /etc/systemd/system/tmp-test.mount
[Unit]
Description=Temporary Directory
Documentation=https://www.freedesktop.org/software/systemd/man/systemd.unit.html
Conflicts=umount.target
Wants=ready-spartan-checker.service
After=ready-spartan-checker.service

[Mount]
What=tmpfs
Where=/tmp/test
Type=tmpfs
Options=mode=1777,strictatime

[Install]
WantedBy=multi-user.target
EOF

sudo systemctl daemon-reload
sudo systemctl enable tmp-test.mount 
sudo reboot
```

When the machine comes back up, you can check the status of the units.
```bash
# Log back  into the machine 
watch -d 'systemctl status tmp-test.mount; systemctl status ready-spartan-checker.service'
```
You will observe that it fails with this error below:
```bash 
ready-spartan-checker.service: Main process exited, code=killed, status=9/KILL
```

**Comment by bernadinm** [comment]
It seems that the flag `TimeoutStartSec=30` it works during the 30 second mark

```bash
ready-spartan-checker.service: Failed with result 'timeout'.
```

But setting it to `TimeoutStartSec=300` it fails during the 60 second mark 

```bash
ready-spartan-checker.service: Failed with result 'signal'.
```

Anyone know how to make it fail properly?

**Comment by poettering** [comment]
Please provide logs for when this fails. Even better provide debug logs. i.e. reproduce the issue and use "systemd-analyze set-log-level debug" first so that debug logging is enabled, then paste the relevant logs here.

**Comment by poettering** [comment]
btw, i am pretty sure you want Type=oneshot for your service that waits, and then use ExecStart= instead of ExecStartPre=

**Comment by bernadinm** [comment]
@poettering, thanks for your feedback. I've also tried Type=oneshot for this and the behavior is the same. I did see that this is not reproducible with the while loop. It is only when you have the loop in the ExecStartPre that the systemd unit fails with `Failed with result 'signal'.` rather than `Failed with result 'timeout'.` which is what it should be. 

When I get some free cycles, I can see if I can provide you the output of `systemd-analyze set-log-level debug` but apart from that, the instructions provided will reproduce this 100% of the time.
